### PR TITLE
Begin Validating Metric Names

### LINKED ISSUE #181: [Bug] Metric names aren't run through name regex
### Housekeeping

- [X] I am a maintainer of dbt-semantic-interfaces

### Short description

Metric names aren't run through the [valid name regex](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/4f6cc55665e5fbe755dfaf8f9bc942020ef15ee5/dbt_semantic_interfaces/validations/unique_valid_name.py#L66) in [UniqueAndValidNameRule](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/4f6cc55665e5fbe755dfaf8f9bc942020ef15ee5/dbt_semantic_interfaces/validations/unique_valid_name.py#L204) even though the rule documents that it is.

### Acceptance criteria

Metric names are validated against the [valid name regex](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/4f6cc55665e5fbe755dfaf8f9bc942020ef15ee5/dbt_semantic_interfaces/validations/unique_valid_name.py#L66)

### Impact to Other Teams

core

### Will backports be required?

0.3.latest

### Context

_No response_

### PR DESCRIPTION
**Pull request by QMalcolm**
Resolves #181

### Description

We've been saying that we require metric names to match the regex (must start with a letter, must be lower case, can't contain dunders, and etc), but we haven't actually been validating this. This PR updates the `UniqueAndValidNameRule` to actually begin validating metric names. It's worth noting that we plan on backporting this to 0.3.latest as not validating the name doesn't cause issues in this repo, but will cause issues in the semantic layer.

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [X] This PR includes tests, or tests are not required/relevant for this PR
- [X] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Comment by QMalcolm** [comment]
@Jstein77 It's worth noting that we may need some messaging around this as people could have specified invalid names in DSI 0.2.x (used by dbt-core 1.6.x). I've checked the [fivetran dbt_add_reporting](https://github.com/fivetran/dbt_ad_reporting) package, and it appears that this change **_won't_** cause any issues for that package.

**Comment by QMalcolm** [comment]
### This is a stacked PR
- https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184 👈 (You are here)
  - https://github.com/dbt-labs/dbt-semantic-interfaces/pull/185

File: tests/validations/test_unique_valid_name.py [hunk 1]
```diff
@@ -33,6 +33,27 @@
...
+    validator = SemanticManifestValidator[PydanticSemanticManifest](
+        [UniqueAndValidNameRule[PydanticSemanticManifest]()]
+    )
+
+    # Shouldn't raise an exception
+    validator.checked_validations(simple_semantic_manifest__with_primary_transforms)
+
+    # Should raise an exception
+    compied_manifest = deepcopy(simple_semantic_manifest__with_primary_transforms)
+    semantic_model = compied_manifest.semantic_models[0]
```
Comment (tlento) [inline tests/validations/test_unique_valid_name.py:46]: nit: `compied` -> `copied`

File: tests/validations/test_unique_valid_name.py [hunk 2]
```diff
@@ -70,6 +70,27 @@ def test_duplicate_semantic_model_name(  # noqa: D
...
+    validator = SemanticManifestValidator[PydanticSemanticManifest](
+        [UniqueAndValidNameRule[PydanticSemanticManifest]()]
+    )
+
+    # Shouldn't raise an exception
+    validator.checked_validations(simple_semantic_manifest__with_primary_transforms)
+
+    # Should raise an exception
+    compied_manifest = deepcopy(simple_semantic_manifest__with_primary_transforms)
+    metric = compied_manifest.metrics[0]
```
Comment (tlento) [inline tests/validations/test_unique_valid_name.py:83]: compied copied compied copied. 😛

**Review by tlento — APPROVED** [review]
Seems reasonable! As discussed let's mark this as a breaking change.

File: .changes/unreleased/Fixes-20231023-103143.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+kind: Fixes
```
[the remark this answers is not in the document]
Comment (QMalcolm) [reply .changes/unreleased/Fixes-20231023-103143.yaml]: Corrected in: [e376244](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/e37624448fed61c44057540fcd08de318adbdafe)
Will be fixed up into: [f0b1389](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/f0b138925841f792c62f83091ef48991f7e40dfd)

File: .changes/unreleased/Fixes-20231023-103143.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+body: Actually validate metric names like we say we do
```
[the remark this answers is not in the document]
Comment (QMalcolm) [reply .changes/unreleased/Fixes-20231023-103143.yaml]: Corrected in: [e376244](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/e37624448fed61c44057540fcd08de318adbdafe)
Will be fixed up into: [f0b1389](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/f0b138925841f792c62f83091ef48991f7e40dfd)

File: tests/validations/test_unique_valid_name.py (hunk 1 above)
Comment (QMalcolm) [reply tests/validations/test_unique_valid_name.py]: Corrected in: [98e21d0](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/98e21d0dd8925475e2cf5f62b7f43cf76265642d)
Will be fixed up into: [bd56473](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/bd564730107d965162dab051fcf7e36724408ac7)

File: tests/validations/test_unique_valid_name.py (hunk 2 above)
Comment (QMalcolm) [reply tests/validations/test_unique_valid_name.py]: Corrected in: [b99cc42](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/b99cc4272bbf6b5f6b54f4d5562016329d7d6eee)
Will be fixed up into: [f0b1389](https://github.com/dbt-labs/dbt-semantic-interfaces/pull/184/commits/f0b138925841f792c62f83091ef48991f7e40dfd)

**Comment by QMalcolm** [comment]
Rebasing to fixup changes into relevant commits as noted in comments

**Pull request merged**